docs/content/configuration/virtualserver-and-virtualserverroute-resources.md

```diff
@@ -95,6 +96,27 @@ basedOn: scheme
 |``basedOn`` | The attribute of a request that NGINX will evaluate to send a redirect. The allowed values are ``scheme`` (the scheme of the request) or ``x-forwarded-proto`` (the ``X-Forwarded-Proto`` header of the request). The default is ``scheme``. | ``string`` | No | ### VirtualServer.Policy |
 {{% /table %}}
 
+### VirtualServer.TLS.CertManager
+
+The cert-manager field configures x509 automated Certificate management for VirtualServer resources using cert-manager (cert-manager.io). Please see the [cert-manager configuration documentation](https://cert-manager.io/docs/configuration/) for more information on deploying and configuring Issuers (Please note that ACME Issuers are not yet supported). Example:
+```yaml
+cert-manager:
+  cluster-issuer: "my-issuer-name"
+```
+
+{{% table %}}
+|Field | Description | Type | Required |
+| ---| ---| ---| --- |
+|``issuer`` |  the name of an Issuer. An Issuer is a cert-manager resource which describes the certificate authority capable of signing certificates. The Issuer must be in the same namespace as the VirtualServer resource. Please note that one of `issuer` and `cluster-issuer` are required, but they are mutually exclusive - one and only one must be defined. | ``string`` | No |
+|``cluster-issuer`` | the name of a ClusterIssuer. A ClusterIssuer is a cert-manager resource which describes the certificate authority capable of signing certificates. It does not matter which namespace your VirtualServer resides, as ClusterIssuers are non-namespaced resources. Please note that one of `issuer` and `cluster-issuer` are required, but they are mutually exclusive - one and only one must be defined. | ``string`` | No |
+|``issuer-kind`` | The kind of the external issuer resource, for example AWSPCAIssuer. This is only necessary for out-of-tree issuers. This cannot be defined if `cluster-issuer` is also defined. | ``string`` | No |
+|``issuer-group`` | The API group of the external issuer controller, for example awspca.cert-manager.io. This is only necessary for out-of-tree issuers. This cannot be defined if `cluster-issuer` is also defined. | ``string`` | No |
+|``common-name`` | This field allows you to configure spec.commonName for the Certificate to be generated. This configuration adds a CN to the x509 certificate. | ``string`` | No |
+|``duration`` | This field allows you to configure spec.duration field for the Certificate to be generated. | ``string`` | No |
+|``renew-before`` |  this annotation allows you to configure spec.renewBefore field for the Certificate to be generated. | ``string`` | No |
+|``usages`` |  This field allows you to configure spec.usages field for the Certificate to be generated. Pass a string with comma-separated values i.e. ``key agreement,digital signature, server auth``. | ``string`` | No |
```

Comment:
Cert-manager documentation lists the valid inputs, we should refer the to full set.

Is there a test that misconfigures these? What's that error and how is it communicated?

Reply:
Yes, we have a test that misconfigures these (in `helper_test.go`). I'll add a reference to the full set.

If misconfigured, it emits a warning and an event, e.g.:

`E0407 15:05:24.206276       1 sync.go:77] Incorrect cert-manager configuration for VirtualServer resource: invalid cert manager field "tls.cert-manager.usages": "something", invalid cert manager field "tls.cert-manager.usages": "funnythat", invalid cert manager field "tls.cert-manager.usages": "nogood"
I0407 15:05:24.206928       1 event.go:285] Event(v1.ObjectReference{Kind:"VirtualServer", Namespace:"default", Name:"cafe", UID:"63d1bd9f-eb3e-46d5-918f-420e3b6d9625", APIVersion:"k8s.nginx.org/v1", ResourceVersion:"86529", FieldPath:""}): type: 'Warning' reason: 'BadConfig' Incorrect cert-manager configuration for VirtualServer resource: invalid cert manager field "tls.cert-manager.usages": "something", invalid cert manager field "tls.cert-manager.usages": "funnythat", invalid cert manager field "tls.cert-manager.usages": "nogood"`